### PR TITLE
feat(explain): add AI-powered checkpoint summaries (ENT-108)

### PR DESCRIPTION
**Pull request by khaong**
## Summary

Adds `--generate` flag to `entire explain -c <checkpoint>` that uses Claude CLI to generate structured AI summaries from checkpoint transcripts.

**New flags:**
- `--generate` - Generate AI summary for a checkpoint
- `--force` - Regenerate even if summary already exists

**Example:**
```bash
entire explain -c abc123 --generate
```

## Changes

- Add `Summary`, `LearningsSummary`, `CodeLearning` types to checkpoint package
- Add `SummaryGenerator` interface for configurable LLM backends
- Implement `ClaudeCLIGenerator` that shells out to `claude` CLI
- Add `BuildCondensedTranscript()` to extract user/assistant/tool entries
- Add `UpdateSummary()` method to persist summaries to `entire/sessions` branch
- Update `formatCheckpointOutput()` to display AI summaries

## Summary Schema

```json
{
  "intent": "What the user wanted to accomplish",
  "outcome": "What was achieved",
  "learnings": {
    "repo": ["Codebase-specific patterns"],
    "code": [{"path": "file.go", "line": 42, "finding": "..."}],
    "workflow": ["General dev practices"]
  },
  "friction": ["Problems encountered"],
  "open_items": ["Tech debt, unfinished work"]
}
```

## Known Limitations

- **Claude-only**: AI summary generation currently only supports Claude Code transcripts. Gemini CLI uses a different transcript format that is not yet supported. See [ENT-144](https://linear.app/entirehq/issue/ENT-144) for planned Gemini support.

## Test plan

- [x] Unit tests for condensed transcript builder
- [x] Unit tests for ClaudeCLIGenerator (with mocked command runner)
- [x] Unit tests for summary display formatting
- [x] Unit tests for UpdateSummary persistence
- [x] Manual test with real Claude CLI

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

> [!NOTE]
> **Medium Risk**
> Adds a new workflow that shells out to the external `claude` CLI and writes new summary commits into the `entire/sessions` metadata branch; failures or unexpected output could impact `explain` UX and repository metadata history.
> 
> **Overview**
> `entire explain --checkpoint` can now **generate and persist an AI summary** via new `--generate`/`--force` flags, storing the result back into checkpoint `metadata.json` on the `entire/sessions` branch.
> 
> Checkpoint metadata gains a structured `Summary` (intent/outcome/learnings/friction/open items), and `explain` output is updated to display these fields (with prompt-based fallbacks when absent).
> 
> Introduces a new `summarise` package that condenses transcripts (user/assistant/tool entries) and a `ClaudeGenerator` that invokes the `claude` CLI and parses its JSON output, plus a new shared `transcript` package to centralize transcript types/parsing utilities and reduce duplication.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 3e8c5d81ad96b72b7c58e49d9ce6a47b52b94650. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: cmd/entire/cli/summarise/summarise.go [hunk 1]
```diff
@@ -0,0 +1,306 @@
...
+}
+
+// toolInput represents the input to various tools
+type toolInput struct {
+	FilePath     string `json:"file_path,omitempty"`
+	NotebookPath string `json:"notebook_path,omitempty"`
+	Description  string `json:"description,omitempty"`
+	Command      string `json:"command,omitempty"`
+	Pattern      string `json:"pattern,omitempty"`
+}
```
[the remark this answers is not in the document]
Comment (khaong) [reply cmd/entire/cli/summarise/summarise.go]: Fixed in ee20b91 - extracted shared types to `cmd/entire/cli/transcript/` package. Both `cli` and `summarise` packages now import from this shared location.

File: cmd/entire/cli/explain.go [hunk 2]
```diff
@@ -201,6 +235,52 @@ func runExplainCheckpoint(w io.Writer, checkpointIDPrefix string, noPager, verbo
...
+		return fmt.Errorf("checkpoint %s has no transcript to summarise", checkpointID.String()[:checkpointIDDisplayLength])
+	}
+
+	// Build condensed transcript for summarisation
+	condensed, err := summarise.BuildCondensedTranscriptFromBytes(result.Transcript)
+	if err != nil {
+		return fmt.Errorf("failed to parse transcript: %w", err)
+	}
+	if len(condensed) == 0 {
+		return fmt.Errorf("checkpoint %s transcript has no content to summarise", checkpointID.String()[:checkpointIDDisplayLength])
```
[the remark this answers is not in the document]
Comment (khaong) [reply cmd/entire/cli/explain.go]: Fixed in ee20b91 - removed the redundant slicing since `CheckpointID.String()` is guaranteed to return exactly 12 characters.

File: cmd/entire/cli/summarise/summarise.go [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+				}
+			}
+		}
+		if len(texts) > 0 {
+			return textutil.StripIDEContextTags(strings.Join(texts, "\n\n"))
+		}
+	}
+
+	return ""
+}
```
[the remark this answers is not in the document]
Comment (khaong) [reply cmd/entire/cli/summarise/summarise.go]: Fixed in ee8003a - moved `extractUserContentFromMessage` to `transcript.ExtractUserContent` in the shared transcript package. Both cli and summarise packages now use this shared implementation.

File: cmd/entire/cli/explain.go [hunk 4]
```diff
@@ -362,11 +442,8 @@ func formatCheckpointOutput(result *checkpoint.ReadCommittedResult, checkpointID
 	meta := result.Metadata
 
 	// Header - always shown
-	shortID := checkpointID
-	if len(shortID) > checkpointIDDisplayLength {
-		shortID = shortID[:checkpointIDDisplayLength]
-	}
-	fmt.Fprintf(&sb, "Checkpoint: %s\n", shortID)
+	// Note: CheckpointID is always exactly 12 characters, matching checkpointIDDisplayLength
```
[the remark this answers is not in the document]
Comment (khaong) [reply cmd/entire/cli/explain.go:447]: The comment is actually correct - it explains WHY we don't need truncation (CheckpointID is always exactly 12 chars). No change needed.

File: cmd/entire/cli/explain.go [hunk 5]
```diff
@@ -102,6 +117,8 @@ Only one of --session, --commit, or --checkpoint can be specified at a time.`,
 	cmd.Flags().BoolVar(&noPagerFlag, "no-pager", false, "Disable pager output")
 	cmd.Flags().BoolVarP(&shortFlag, "short", "s", false, "Show summary only (omit prompts and files)")
 	cmd.Flags().BoolVar(&fullFlag, "full", false, "Show complete transcript")
+	cmd.Flags().BoolVar(&generateFlag, "generate", false, "Generate an AI summary for the checkpoint")
+	cmd.Flags().BoolVar(&forceFlag, "force", false, "Regenerate summary even if one already exists (requires --generate)")
```
[the remark this answers is not in the document]
Comment (khaong) [reply cmd/entire/cli/explain.go:122]: This was already present at line 124 after the merge - no action needed.

File: cmd/entire/cli/explain.go [hunk 6]
```diff
@@ -200,6 +222,18 @@ func runExplainCheckpoint(w io.Writer, checkpointIDPrefix string, noPager, verbo
...
+	// Handle summary generation
+	if generate {
+		if err := generateCheckpointSummary(w, errW, store, fullCheckpointID, result, force); err != nil {
+			return err
+		}
+		// Reload the result to get the updated summary
+		result, err = store.ReadCommitted(context.Background(), fullCheckpointID)
+		if err != nil {
+			return fmt.Errorf("failed to reload checkpoint: %w", err)
+		}
```
[the remark this answers is not in the document]
Comment (khaong) [reply cmd/entire/cli/explain.go:234]: Fixed in ee8003a - added nil check after reloading the checkpoint result.

File: cmd/entire/cli/summarise/summarise.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (khaong) [reply cmd/entire/cli/summarise/summarise.go]: Fixed in ee8003a - moved `extractUserContentFromMessage` to `transcript.ExtractUserContent` in the shared transcript package. Both cli and summarise packages now use this shared implementation.

File: cmd/entire/cli/summarise/summarise.go [hunk 7]
```diff
@@ -0,0 +1,267 @@
...
+			lines = append(lines, line)
+		}
+
+		if err == io.EOF {
+			break
+		}
+	}
+
+	return lines, nil
+}
```
[the remark this answers is not in the document]
Comment (khaong) [reply cmd/entire/cli/summarise/summarise.go]: Fixed in ee8003a - moved `parseTranscriptFromBytes` to `transcript.ParseFromBytes` in the shared transcript package. Both cli and summarise packages now use this shared implementation.

File: cmd/entire/cli/summarise/summarise.go [hunk 8]
```diff
@@ -0,0 +1,267 @@
...
+	"entire.io/cli/cmd/entire/cli/textutil"
+	"entire.io/cli/cmd/entire/cli/transcript"
+)
+
+// Generator generates checkpoint summaries using an LLM.
+type Generator interface {
+	// Generate creates a summary from checkpoint data.
+	// Returns the generated summary or an error if generation fails.
+	Generate(ctx context.Context, input Input) (*checkpoint.Summary, error)
+}
```
[the remark this answers is not in the document]
Comment (khaong) [reply cmd/entire/cli/summarise/summarise.go:19]: Fixed in ee8003a - changed `explain.go` to use the interface type: `var generator summarise.Generator = &summarise.ClaudeGenerator{}`. This demonstrates the interface is intentionally designed for testability and extensibility.

File: cmd/entire/cli/summarise/summarise.go [hunk 9]
```diff
@@ -0,0 +1,195 @@
...
+
+	// ToolName is the name of the tool (for tool entries)
+	ToolName string
+
+	// ToolDetail is a description or file path (for tool entries)
+	ToolDetail string
+}
+
+// TranscriptLine is an alias to the shared transcript.Line type.
+type TranscriptLine = transcript.Line
```
[the remark this answers is not in the document]
Comment (khaong) [reply cmd/entire/cli/summarise/summarise.go]: Fixed in 3f1eae5 - removed the `TranscriptLine` type alias from the summarise package. Now using `transcript.Line` directly. Note: the alias in `claudecode/transcript.go` predates this PR and is outside the scope of this change.

File: cmd/entire/cli/transcript.go [hunk 10]
```diff
@@ -219,34 +197,15 @@ func extractUserPromptAt(transcript []transcriptLine, idx int) string {
...
+	return extractUserContentFromMessage(transcript[idx].Message)
+}
 
-	return ""
+// extractUserContentFromMessage extracts user content from a raw message.
+// Handles both string and array content formats.
+// IDE-injected context tags (like <ide_opened_file>) are stripped from the result.
+// Returns empty string if the message cannot be parsed or contains no text.
+func extractUserContentFromMessage(message json.RawMessage) string {
+	return transcript.ExtractUserContent(message)
```
[the remark this answers is not in the document]
Comment (khaong) [reply cmd/entire/cli/transcript.go]: Fixed in 3f1eae5 - inlined the wrapper by calling `transcript.ExtractUserContent()` directly in `extractUserPromptAt`.

File: cmd/entire/cli/transcript/parse.go [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+				}
+			}
+		}
+		if len(texts) > 0 {
+			return textutil.StripIDEContextTags(strings.Join(texts, "\n\n"))
+		}
+	}
+
+	return ""
+}
```
[the remark this answers is not in the document]
Comment (khaong) [reply cmd/entire/cli/transcript/parse.go:80]: Fixed in 188d883 - added unit tests for the transcript package covering `ParseFromBytes` and `ExtractUserContent` with various edge cases (empty content, malformed JSON, IDE tag stripping, tool results).

File: cmd/entire/cli/explain.go [hunk 12]
```diff
@@ -209,6 +246,52 @@ func runExplainCheckpoint(w io.Writer, checkpointIDPrefix string, noPager, verbo
...
+		return fmt.Errorf("checkpoint %s transcript has no content to summarise", checkpointID)
+	}
+
+	input := summarise.Input{
+		Transcript:   condensed,
+		FilesTouched: result.Metadata.FilesTouched,
+	}
+
+	// Show progress indicator
+	fmt.Fprintln(errW, "Generating summary...")
```
[the remark this answers is not in the document]
Comment (khaong) [reply cmd/entire/cli/explain.go]: Fixed in 188d883 - changed to use `logging.Info(ctx, "generating checkpoint summary")` instead of `fmt.Fprintln` to stderr.

File: cmd/entire/cli/transcript/parse.go [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+			lines = append(lines, line)
+		}
+
+		if err == io.EOF {
+			break
+		}
+	}
+
+	return lines, nil
+}
```
[the remark this answers is not in the document]
Comment (khaong) [reply cmd/entire/cli/transcript/parse.go:45]: The `claudecode.ParseTranscript` function predates this PR and is used in different contexts (with a larger buffer for file-based parsing). Unifying these implementations would require changes to the claudecode package which is out of scope for this PR. Created as a follow-up consideration.

File: cmd/entire/cli/agent/claudecode/transcript.go [hunk 14]
```diff
@@ -10,40 +10,18 @@ import (
...
-	Text  string          `json:"text,omitempty"`
-	Name  string          `json:"name,omitempty"`
-	Input json.RawMessage `json:"input,omitempty"`
-}
-
-// toolInput represents the input to file modification tools
-type toolInput struct {
-	FilePath     string `json:"file_path,omitempty"`
-	NotebookPath string `json:"notebook_path,omitempty"`
-}
```
[the remark this answers is not in the document]
Comment (khaong) [reply cmd/entire/cli/agent/claudecode/transcript.go:46]: The `claudecode.ExtractLastUserPrompt` function predates this PR and has different behavior (finds the last user prompt, doesn't strip IDE tags). The shared `transcript.ExtractUserContent` was created to consolidate the extraction logic used by the cli and summarise packages. Unifying with claudecode would be a larger refactoring effort outside the scope of this PR.

File: cmd/entire/cli/explain.go [hunk 15]
```diff
@@ -209,6 +246,51 @@ func runExplainCheckpoint(w io.Writer, checkpointIDPrefix string, noPager, verbo
...
+	if err != nil {
+		return fmt.Errorf("failed to generate summary: %w", err)
+	}
+
+	// Persist the summary
+	if err := store.UpdateSummary(ctx, checkpointID, summary); err != nil {
+		return fmt.Errorf("failed to save summary: %w", err)
+	}
+
+	fmt.Fprintln(w, "Summary generated and saved.")
```
[the remark this answers is not in the document]
Comment (khaong) [reply cmd/entire/cli/explain.go]: Fixed in 3e8c5d8 - changed to use "✓ Summary generated and saved" to align with the established codebase pattern (e.g., setup.go).

File: cmd/entire/cli/summarise/claude.go [hunk 16]
```diff
@@ -0,0 +1,174 @@
...
+		return nil, fmt.Errorf("failed to parse summary JSON: %w (response: %s)", err, resultJSON)
+	}
+
+	return &summary, nil
+}
+
+// BuildSummarisationPrompt creates the prompt for the Claude CLI.
+func BuildSummarisationPrompt(transcriptText string) string {
+	return fmt.Sprintf(summarisationPromptTemplate, transcriptText)
+}
```
[the remark this answers is not in the document]
Comment (khaong) [reply cmd/entire/cli/summarise/claude.go:148]: This is already unexported - the function is named `buildSummarisationPrompt` (lowercase). The comment appears to be based on outdated analysis.

[the remark this answers is not in the document]
Comment (khaong) [reply cmd/entire/cli/summarise/claude.go:148]: fixed in https://github.com/entireio/cli/pull/121/commits/2886ea57cd17b1e94351ec56937e564b1252b0b6

File: cmd/entire/cli/explain.go [hunk 17]
```diff
@@ -209,6 +246,51 @@ func runExplainCheckpoint(w io.Writer, checkpointIDPrefix string, noPager, verbo
 	return nil
 }
 
+// generateCheckpointSummary generates an AI summary for a checkpoint and persists it.
+func generateCheckpointSummary(w, _ io.Writer, store *checkpoint.GitStore, checkpointID id.CheckpointID, result *checkpoint.ReadCommittedResult, force bool) error {
```
[the remark this answers is not in the document]
Comment (khaong) [reply cmd/entire/cli/explain.go:250]: minor nit
